lib/cartopy/mpl/gridliner.py

```diff
@@ -511,23 +502,34 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None,
                     elif isinstance(intersection,
                                     sgeom.collection.GeometryCollection):
                         # This is a collection of Point and LineString that
-                        # represent the same gridline. We only consider
-                        # the first and last geometries, merge their
-                        # coordinates and keep first or last two points
-                        # to get only one tail and and one head.
-                        tails = []
-                        heads = []
-                        for ht, slicer in [(tails, slice(0, 2)),
-                                           (heads, slice(-1, -3, -1))]:
-                            x = array('d', [])
-                            y = array('d', [])
-                            for geom in list(intersection.geoms)[slicer]:
-                                x += geom.xy[0]
-                                y += geom.xy[1]
-                            ht.append(list(zip(x[slicer], y[slicer])))
```

Comment:
@stefraynaud I think there used to be a bug here, but can you confirm?

When producing `heads`, it looks at the last geom, grabs its coordinates (in order), then appends the coordinates (in order) of the second-last geom. Finally, it takes the last two points from that list of coordinates in reverse. If the last two geoms are `Point`s, then you get the second-last and then the last coordinate.

But if the second-last geom is a `LineString`, then you get the last two coordinates of the second-last geom, when it should be the last coordinate of the second-last geom and the coordinate of the last geom.

Also, for the new code, I'm not sure if it `heads` should be in order, or reversed.

Reply:
Here is an example:
```python
In [6]: def new(intersection): 
   ...:     xy = [] 
   ...:     for geom in intersection.geoms: 
   ...:         for coord in geom.coords: 
   ...:             xy.append(coord) 
   ...:             if len(xy) == 2: 
   ...:                 break 
   ...:         if len(xy) == 2: 
   ...:             break 
   ...:     tails = [xy] 
   ...:     xy = [] 
   ...:     for geom in reversed(intersection.geoms): 
   ...:         for coord in reversed(geom.coords): 
   ...:             xy.append(coord) 
   ...:             if len(xy) == 2: 
   ...:                 break 
   ...:         if len(xy) == 2: 
   ...:             break 
   ...:     heads = [xy] 
   ...:     return tails, heads 
   ...:                                                                                                                                               

In [7]: def orig(intersection): 
   ...:     tails = [] 
   ...:     heads = [] 
   ...:     for ht, slicer in [(tails, slice(0, 2)), 
   ...:                        (heads, slice(-1, -3, -1))]: 
   ...:         x = array('d', []) 
   ...:         y = array('d', []) 
   ...:         for geom in list(intersection.geoms)[slicer]: 
   ...:             x += geom.xy[0] 
   ...:             y += geom.xy[1] 
   ...:         ht.append(list(zip(x[slicer], y[slicer]))) 
   ...:     return tails, heads 
   ...:                                                                                                                                               
```

If it's just `Point`s, then the result is:
```python
In [8]: g = sgeom.GeometryCollection([sgeom.Point(i, i) for i in range(10)])                                                                          

In [9]: orig(g)                                                                                                                                       
Out[9]: ([[(0.0, 0.0), (1.0, 1.0)]], [[(8.0, 8.0), (9.0, 9.0)]])

In [10]: new(g)                                                                                                                                       
Out[10]: ([[(0.0, 0.0), (1.0, 1.0)]], [[(9.0, 9.0), (8.0, 8.0)]])
```
So I should flip the `heads` to be in order. I'm not sure if it's semantically correct, but at least it will match.

But if you use `LineString`s (`x` is the geometry, and `y` is the point in the geometry):
```python
In [11]: g = sgeom.GeometryCollection([sgeom.LineString([(x, y) for y in range(10)]) for x in range(10)])                                             

In [12]: g                                                                                                                                            
Out[12]: <shapely.geometry.collection.GeometryCollection at 0x7f68e544b550>

In [13]: orig(g)                                                                                                                                      
Out[13]: ([[(0.0, 0.0), (0.0, 1.0)]], [[(8.0, 9.0), (8.0, 8.0)]])

In [14]: new(g)                                                                                                                                       
Out[14]: ([[(0.0, 0.0), (0.0, 1.0)]], [[(9.0, 9.0), (9.0, 8.0)]])
```
So the original code picks the second-last geometry and returns the last two elements from it. Not only that, but this time it returns the elements in reverse.

Reply:
This code path is only triggered 5 times in `test_grid_labels_inline`, and in all cases, it's 10% `Point`, followed by `LineString` (except the one time it's 1 of each.) So `tails` is always consistent between old and new, but `heads` is always reversed in the code as it stands. Thus I will leave it reversed for the `Point`s case even though the original didn't do that.

Reply:
It seems you're right